### ISSUE TITLE
Shapely 1.8 array interface deprecation for object arrays

### ISSUE THREAD

**Issue by mwtoews** [Toblerity/Shapely#1096]
I'm looking at some of the shapely 1.8 deprecation impacts in a niche package, and I'm stumbling over one issue that can be represented in a small example:

```python
import numpy as np
from shapely.geometry import Point, MultiPoint
mp = MultiPoint([Point(1, 2), Point(3, 4)])
mp_list = [mp]
a = np.empty(len(mp_list), dtype=object)
a[:] = mp_list
print(a[0].wkt)
```
This example works for all current and development versions of shapely, printing `MULTIPOINT (1 2, 3 4)` from the object array.
 
Shapely 1.7 does not show any warnings.

Shapely 1.8dev shows two warnings:
> /usr/bin/ipython3:1: ShapelyDeprecationWarning: `__len__` for multi-part geometries is deprecated and will be removed in Shapely 2.0. Check the length of the `geoms` property instead to get the  number of parts of a multi-part geometry.
  #! /bin/sh
/usr/lib/python3/dist-packages/IPython/core/interactiveshell.py:2882: ShapelyDeprecationWarning: The array interface is deprecated and will no longer work in Shapely 2.0. Convert the '.coords' to a numpy array instead.
  exec(code_obj, self.user_global_ns, self.user_ns)

Shapely 2.0dev does not show any warnings, and works the same as any other version.

---

It seems that the warnings from shapely 1.8dev are crying wolf. I understand these shenanigans are permitted with object arrays. Is there any way for shapely 1.8 to omit these warnings? Or should I just refactor to ether of these?
```python
a = np.empty(len(mp_list), dtype=object)
for idx, obj in enumerate(mp_list):
    a[idx] = obj

# or

import warnings
with warnings.catch_warnings():
    warnings.simplefilter("ignore")
    a = np.empty(len(mp_list), dtype=object)
    a[:] = mp_list
```

**Comment by jorisvandenbossche** [comment]
Yeah, this is an annoying aspect when dealing with numpy arrays. I encountered the same issue in GeoPandas, and wrote a helper for the "ignore warnings" approach, see https://github.com/geopandas/geopandas/pull/1659 (we do this a lot in geopandas ..)

I am not fully sure why it happens (because in the end the geometries are not converted to its parts or coordinates), but apparently, the warnings are still triggered somewhere in the numpy internals, outside our control. 

Unfortunately, I think the only thing we can do is clearly documenting this case and how to solve it in the release notes for 1.8.

**Comment by mwtoews** [comment]
Thanks for the details @jorisvandenbossche, I was suspecting something was happening deep within the numpy internals. The helper looks good! Here is my adaptation for it, with a few minor edits:
```python
import contextlib
import shapely
import warnings
from distutils.version import LooseVersion

SHAPELY_GE_20 = str(shapely.__version__) >= LooseVersion("2.0")

try:
    from shapely.errors import ShapelyDeprecationWarning as shapely_warning
except ImportError:
    shapely_warning = None

if shapely_warning is not None and not SHAPELY_GE_20:
    @contextlib.contextmanager
    def ignore_shapely2_warnings():
        with warnings.catch_warnings():
            warnings.filterwarnings("ignore", category=shapely_warning)
            yield
else:
    @contextlib.contextmanager
    def ignore_shapely2_warnings():
        yield
```
then to use it:
```python
with ignore_shapely2_warnings():
    a[:] = mp_list
```

**Comment by jorisvandenbossche** [comment]
Yes, in the final version I expanded the captured messages (https://github.com/geopandas/geopandas/blob/8ddac2f9a44870aeeca8bc73dd74efddfce23ad1/geopandas/_compat.py#L146), but just generally capturing `ShapelyDeprecationWarning` is of course fine as well.

Users should be careful to only use that for such cases as the above, and not to silence "real" warnings, though.

**Comment by sgillies** [comment]
Is this issue resolved? If I read it right, the gist of it is that numpy accesses either `__len__` or `__array_interface__` on shapely 1.8 and that the warnings go away with shapely 2.0 because these are gone and numpy succeeds by other paths. Is that right?

**Comment by mwtoews** [comment]
It's resolved from my perspective, and only applies to 1.8.

But other users may run into this too, so I'm wondering if this is worth documenting somewhere, outside of this issue.

**Comment by jorisvandenbossche** [comment]
Yes, I think we should document this (it's still on my todo to write an expanded 1.8 whatsnew explaining the different deprecations ...)

**Comment by Rikuoja** [comment]
Happy I found this thread as way of documentation, btw!

My issue is that Pandas itself accesses an object array in https://github.com/pandas-dev/pandas/blob/master/pandas/core/dtypes/cast.py#L1996 . In this case, if values is a GeoDataframe with a Shapely 1.8 object, this will crash Pandas with

```
/opt/conda/lib/python3.8/site-packages/pandas/core/dtypes/cast.py:118: ShapelyDeprecationWarning: The array interface is deprecated and will no longer work in Shapely 2.0. Convert the '.coords' to a numpy array instead.
  arr = construct_1d_object_array_from_listlike(values)
```

I guess this should be considered a Shapely issue, or should there be a bug report in GeoPandas too?

**Comment by jorisvandenbossche** [comment]
FYI this should also be documented at https://shapely.readthedocs.io/en/latest/migration.html#creating-numpy-arrays-of-geometry-objects now

> My issue is that Pandas itself accesses an object array in https://github.com/pandas-dev/pandas/blob/master/pandas/core/dtypes/cast.py#L1996 . In this case, if values is a GeoDataframe with a Shapely 1.8 object, this will crash Pandas core with

Note that it is not a "crash", but only a warning (not that warnings can't be annoying of course). And depending on your use case (can you show an example code?), the behaviour might not actually change in the future and you could ignore the warning.

In GeoPandas, that's actually what we do in certain places where pandas generates the warning and so we can't avoid it, then we suppress the warning for our users. See eg https://github.com/geopandas/geopandas/blob/733e43f4cf9dc140130e4fbdcc3ff17ba43905e8/geopandas/geoseries.py#L192-L199

**Comment by jorisvandenbossche** [comment]
I notice your updated comment now:

> or should there be a bug report in GeoPandas too, i.e. have you missed catching this warning in Geopandas? I am running Geopandas 0.10.2 which should have merged capturing the warnings?

Yes, an issue would certainly be welcome if you have a reproducible code example that still gives a warning through geopandas (as I think the goal at least is to catch those warnings as explained above)

**Comment by Rikuoja** [comment]
Good point. My problem is that I am trying out a notebook with external functions that use pandas, and I couldn't get my notebook to catch the full stack trace, so I haven't found the offending line causing the call in pandas. I will try to get a code example and get back to you!

**Comment by Rikuoja** [comment]
Okay, narrowed it down to a pure pandas call that makes a Shapely point out of a coordinate array, like below

```
 results['geometry']=[Point(geom[1], geom[0]) for geom in results['geometry']]
```

This will of course call `construct_1d_object_array_from_listlike` in Pandas. Therefore, it is not a Geopandas issue, but a Shapely issue. Thanks for your help again!

**Comment by jorisvandenbossche** [comment]
If you want to avoid the warning (without doing a filterwarnings yourself), you can also convert the list to a GeoSeries before assigning it to the column of the dataframe:

```
results["geometry"] = geopandas.GeoSeries([..list comprehension..], index=results.index)
```

or otherwise specify this while converting to a GeoDataFrame (`geopandas.GeoDataFrame(results, geometry=[..list comprehension..])`)

**Comment by Rikuoja** [comment]
Ah, of course! Thank you for the suggestion.
